### PR TITLE
Adjust parent types for exceptions

### PR DESCRIPTION
**Pull request by idelpivnitskiy**
Motivation:

After revisiting all existing exception types, we found that some of
them use incorrect parent type.

Modifications:

- Use more appropriate parent type for existing exceptions:
  - `ConnectionRejectedException` extend `RuntimeException` -> `SocketException`;
  - `NoAvailableHostException` extend `RuntimeException` -> `IOException`;
  - `DuplicateAttributeException` extend `RuntimeException` -> `IllegalStateException`;
  - `TerminateRepeatException` extend `Exception` -> `RuntimeException`;
  - `ClosedServiceDiscovererException` extend `RuntimeException` -> `ClosedChannelException`;
  - `UnsupportedHttpChunkException` extend `SerializationException` -> `IllegalArgumentException`;
- Regenerate `serialVersionUID` for updated exception classes;
- Use `StacklessConnectionRejectedException` for
`RoundRobinLoadBalancer`;
- Mark `MessageEncodingException` as `@Deprecated` (was missed during rework of 
serialization API);
- Remove unused ctor from `ResponseTimeoutTest`;

Result:

Exceptions have more appropriate parent types.

### CODE REVIEW

File: servicetalk-client-api/src/main/java/io/servicetalk/client/api/ConnectionRejectedException.java [hunk 1]
```diff
@@ -17,11 +17,13 @@
...
 import io.servicetalk.transport.api.RetryableException;
 
+import java.net.SocketException;
+
 /**
  * Thrown when a newly created connection was rejected.
  */
-public final class ConnectionRejectedException extends RuntimeException implements RetryableException {
-    private static final long serialVersionUID = -2573006397795678870L;
+public class ConnectionRejectedException extends SocketException implements RetryableException {
```
Comment (idelpivnitskiy) [inline servicetalk-client-api/src/main/java/io/servicetalk/client/api/ConnectionRejectedException.java:25]: Debated between `ConnectException` and `SocketException`. Based on their javadoc, decided to go with `SocketException` because of technically the connection was already established, but there was a problem "accessing" it due to a selector or a host becoming unavailable.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RepeatStrategies.java [hunk 2]
```diff
@@ -40,8 +40,8 @@
     /**
      * An {@link Exception} instance used to indicate termination of repeats.
      */
-    public static final class TerminateRepeatException extends Exception {
-        private static final long serialVersionUID = -1725458427890873886L;
+    public static final class TerminateRepeatException extends RuntimeException {
```
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RepeatStrategies.java:43]: The decision to throw this exception happens at runtime

File: servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/MessageEncodingException.java [hunk 3]
```diff
@@ -15,10 +15,18 @@
...
+
 /**
  * Exception thrown when a message was encoded with an unsupported encoder.
+ *
+ * @deprecated This exception type was thrown only by {@code ProtoBufSerializationProviderBuilder} that was deprecated
+ * and will be removed in future releases. Use {@link SerializationException} instead that can be thrown by a new
+ * implementation.
  */
+@Deprecated
 public final class MessageEncodingException extends RuntimeException {
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-api/src/main/java/io/servicetalk/grpc/api/MessageEncodingException.java:28]: `ProtoBufSerializationProviderBuilder` was deprecated only in the main branch, we don't need to cherry-pick this into 0.41

File: servicetalk-grpc-protobuf/src/main/java/io/servicetalk/grpc/protobuf/ProtoBufSerializationProviderBuilder.java [hunk 4]
```diff
@@ -172,7 +172,7 @@ public GrpcSerializationProvider build() {
             @SuppressWarnings("unchecked")
             HttpSerializer<T> httpSerializer = serializersForType.get(codec);
             if (httpSerializer == null) {
-                throw new MessageEncodingException("Unknown encoding: " + codec.name());
+                throw new MessageEncodingException(codec.name().toString());
```
Comment (idelpivnitskiy) [inline servicetalk-grpc-protobuf/src/main/java/io/servicetalk/grpc/protobuf/ProtoBufSerializationProviderBuilder.java:175]: Full message is generated inside a ctor

File: servicetalk-http-api/src/main/java/io/servicetalk/http/api/UnsupportedHttpChunkException.java [hunk 5]
```diff
@@ -15,11 +15,9 @@
  */
 package io.servicetalk.http.api;
 
-import io.servicetalk.serializer.api.SerializationException;
-
 import javax.annotation.Nullable;
 
-final class UnsupportedHttpChunkException extends SerializationException {
+final class UnsupportedHttpChunkException extends IllegalArgumentException {
```
Comment (idelpivnitskiy) [inline servicetalk-http-api/src/main/java/io/servicetalk/http/api/UnsupportedHttpChunkException.java:20]: The change from `IllegalArgumentException` to `SerializationException` was made only in the main branch in #1673. This is not related to serialization. Reverting back to `IllegalArgumentException`.
cc @Scottmitch

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java [hunk 6]
```diff
@@ -263,21 +265,32 @@ public StreamingHttpClient apply(final UrlKey urlKey) {
...
         }
 
         @Override
         public Single<StreamingHttpResponse> request(final HttpExecutionStrategy strategy,
                                                      final StreamingHttpRequest request) {
-            return defer(() -> selectClient(request).request(strategy, request).subscribeShareContext());
+            return defer(() -> {
+                try {
+                    return selectClient(request).request(strategy, request).subscribeShareContext();
+                } catch (Throwable t) {
```
Comment (idelpivnitskiy) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:290]: The `try-catch` here is necessary anyway, because methods that return async sources are not expected to throw.

**Review by idelpivnitskiy — COMMENTED** [review]
I don't expect it to be an impactful breaking change because most of the users use our types instead of parent types. Should be an appropriate change for 0.42 release, wdyt?

File: servicetalk-grpc-protobuf/src/main/java/io/servicetalk/grpc/protobuf/ProtoBufSerializationProviderBuilder.java (hunk 4 above)
Comment (idelpivnitskiy) [reply servicetalk-grpc-protobuf/src/main/java/io/servicetalk/grpc/protobuf/ProtoBufSerializationProviderBuilder.java:175]: Decided to open a separate PR #1940 because we should also cherry-pick this into 0.41

File: servicetalk-client-api/src/main/java/io/servicetalk/client/api/NoAvailableHostException.java [hunk 7]
```diff
@@ -17,11 +17,13 @@
...
 import io.servicetalk.transport.api.RetryableException;
 
+import java.net.SocketException;
+
 /**
  * Thrown when no host is available but at least one is required.
  */
-public class NoAvailableHostException extends RuntimeException implements RetryableException {
-    private static final long serialVersionUID = 5340791072245425967L;
+public class NoAvailableHostException extends SocketException implements RetryableException {
```
Comment (tkountis) [inline servicetalk-client-api/src/main/java/io/servicetalk/client/api/NoAvailableHostException.java:25]: Is it a "SocketException"? I would rather keep it a runtime one or associate with a LB/SD parent type.

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java [hunk 8]
```diff
@@ -63,6 +64,9 @@
  * A builder of {@link StreamingHttpClient} instances which have a capacity to call any server based on the parsed
  * absolute-form URL address information from each {@link StreamingHttpRequest}.
  * <p>
+ * If {@link HttpRequestMetaData#requestTarget()} is not an absolute-form URL, a {@link MalformedURLException} will be
```
Comment (bondolo) [inline servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java:67]: Possibly just a personal preference but I usually prefer `URISyntaxException` to `MalformedURLException` because the later often gets handled as a generic `IOException` by lazy people who hate checked exceptions.

**Review by bondolo — APPROVED** [review]

File: servicetalk-client-api/src/main/java/io/servicetalk/client/api/NoAvailableHostException.java (hunk 7 above)
Comment (idelpivnitskiy) [reply servicetalk-client-api/src/main/java/io/servicetalk/client/api/NoAvailableHostException.java:25]: The intend was to make it `IOException` because it prevents a request execution due to unavailability to execute an IO operation. During offline discussion, @bondolo recommended a more specific subtype - `SocketException`. I'm agreed with that proposal because it notifies that there are no sockets available at this point of time.

Comment (tkountis) [reply servicetalk-client-api/src/main/java/io/servicetalk/client/api/NoAvailableHostException.java:25]: Don't see a reason to block the change but I don't agree that this is a good correlation. I agree that `IOException` is a valid super type, but `SocketException` is clearly doc'ed to do with failure to access or create a socket. In our case this is not even related to a connection but rather to host-finding.

Comment (idelpivnitskiy) [reply servicetalk-client-api/src/main/java/io/servicetalk/client/api/NoAvailableHostException.java]: Ok, I was reading "failure to access a socket" as "no hosts -> no access" 😄 
Changing to `IOException`: https://github.com/apple/servicetalk/pull/1924/commits/c2b41cf0f43e5aaf3ce8e7e4e4cbd8c796ef530f

File: servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java (hunk 8 above)
Comment (idelpivnitskiy) [reply servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/DefaultMultiAddressUrlHttpClientBuilder.java]: I've opened a separate PR #1955 for this change to make it move visible in release notes
